README.rst
proc: Linux process information interface
=========================================

.. image:: https://travis-ci.org/xolox/python-proc.svg?branch=master
   :target: https://travis-ci.org/xolox/python-proc

.. image:: https://coveralls.io/repos/xolox/python-proc/badge.svg?branch=master
   :target: https://coveralls.io/r/xolox/python-proc?branch=master

The Python package `proc` exposes process information available in the Linux
`process information pseudo-file system`_ available at ``/proc``. The `proc`
package is currently tested on cPython 2.7, 3.5+ and PyPy (2.7). The automated
test suite regularly runs on Ubuntu Linux but other Linux variants (also those
not based on Debian Linux) should work fine. For usage instructions please
refer to the documentation_.

.. contents::
   :local:

Installation
------------

The `proc` package is available on PyPI_ which means installation should be as
simple as:

.. code-block:: console

   $ pip install proc

There's actually a multitude of ways to install Python packages (e.g. the `per
user site-packages directory`_, `virtual environments`_ or just installing
system wide) and I have no intention of getting into that discussion here, so
if this intimidates you then read up on your options before returning to these
instructions ;-).

Once you've installed the `proc` package head over to the documentation_ for
some examples of how the `proc` package can be used.

Design choices
--------------

The `proc` package was created with the following considerations in mind:

**Completely specialized to Linux**
 It parses ``/proc`` and nothing else ;-).

**Fully implemented in Python**
 No binary/compiled components, as opposed to psutil_ which is way more
 portable but requires a compiler for installation.

**Very well documented**
 The documentation should make it easy to get started (as opposed to procfs_
 which I evaluated and eventually gave up on because I had to resort to reading
 through its source code just to be disappointed in its implementation).

**Robust implementation**
 Reading ``/proc`` is inherently sensitive to race conditions and the `proc`
 package takes this into account, in fact the test suite contains a test that
 creates race conditions in order to verify that they are handled correctly.
 The API of the `proc` package hides race conditions as much as possible and
 where this is not possible the consequences are clearly documented.

**Layered API design** (where each layer is documented)
 Builds higher level abstractions on top of lower level abstractions:

 **The proc.unix module**
  Defines a simple process class that combines process IDs and common UNIX
  signals to implement process control primitives like waiting for a process to
  end and gracefully or forcefully terminating a process.

 **The proc.core module**
  Builds on top of the ``proc.unix`` module to provide a simple, fast and easy
  to use API for the process information available in ``/proc``. If you're
  looking for a simple and/or fast interface to ``/proc`` that does the heavy
  lifting (parsing) for you then this is what you're looking for.

 **The proc.tree module**
  Builds on top of the ``proc.core`` module to provide an in-memory tree data
  structure that mimics the actual process tree, enabling easy searching and
  navigation through the process tree.

 **The proc.apache module**
  Builds on top of the ``proc.tree`` module to implement an easy to use Python
  API that does metrics collection for monitoring of Apache web server worker
  memory usage, including support for WSGI process groups.

 **The proc.cron module**
  Implements the command line program ``cron-graceful`` which gracefully
  terminates cron daemons. This module builds on top of the ``proc.tree``
  module as a demonstration of the possibilities of the `proc` package and as a
  practical tool that is ready to be used on any Linux system that has Python
  and cron_ installed.

 **The proc.notify module**
  Implements the command line program ``notify-send-headless`` which can be
  used to run the program ``notify-send`` in headless environments like cron
  jobs and system daemons.

History
-------

I've been writing shell and Python scripts that parse ``/proc`` for years now
(it seems so temptingly easy when you get started ;-). Sometimes I resorted to
copy/pasting snippets of Python code between personal and work projects because
the code was basically done, just not available in an easy to share form.

Once I started fixing bugs in diverging copies of that code I decided it was
time to combine all of the features I'd grown to appreciate into a single well
tested and well documented Python package with an easy to use API and share it
with the world.

This means that, although I made my first commit on the `proc` package in March
2015, much of its code has existed for years in various forms.

Similar projects
----------------

Below are several other Python libraries that expose process information. If
the `proc` package isn't working out for you consider trying one of these. The
summaries are copied and/or paraphrased from the documentation of each
package:

psutil_
  A cross-platform library for retrieving information on running processes and
  system utilization (CPU, memory, disks, network) in Python.

procpy_
  A Python wrapper for the procps_ library and a module containing higher level
  classes (with some extensions compared to procps_).

procfs_
  Python API for the Linux ``/proc`` virtual filesystem.

Contact
-------

The latest version of `proc` is available on PyPI_ and GitHub_. The
documentation is hosted on `Read the Docs`_ and includes a changelog_. For bug
reports please create an issue on GitHub_. If you have questions, suggestions,
etc. feel free to send me an e-mail at `peter@peterodding.com`_.

License
-------

This software is licensed under the `MIT license`_.

© 2020 Peter Odding.

.. External references:
.. _changelog: https://proc.readthedocs.io/en/latest/changelog.html
.. _cron: http://en.wikipedia.org/wiki/Cron
.. _documentation: https://proc.readthedocs.io
.. _GitHub: https://github.com/xolox/python-proc
.. _MIT license: http://en.wikipedia.org/wiki/MIT_License
.. _per user site-packages directory: https://www.python.org/dev/peps/pep-0370/
.. _peter@peterodding.com: peter@peterodding.com
.. _process information pseudo-file system: http://linux.die.net/man/5/proc
.. _procfs: https://pypi.python.org/pypi/procfs
.. _procps: http://procps.sourceforge.net/
.. _procpy: http://code.google.com/p/procpy/
.. _psutil: https://pypi.python.org/pypi/psutil/
.. _PyPI: https://pypi.python.org/pypi/proc
.. _Read the Docs: https://proc.readthedocs.io
.. _virtual environments: http://docs.python-guide.org/en/latest/dev/virtualenvs/


CHANGELOG.rst
Changelog
=========

The purpose of this document is to list all of the notable changes to this
project. The format was inspired by `Keep a Changelog`_. This project adheres
to `semantic versioning`_.

.. contents::
   :local:

.. _Keep a Changelog: http://keepachangelog.com/
.. _semantic versioning: http://semver.org/

`Release 1.0`_ (2020-04-26)
---------------------------

- Merged `pull request #2`_ which enables Python 3.7+ compatibility by changing
  the :pypi:`executor` integration to stop using the old ``async`` option and
  start using the new ``asynchronous`` option (given that :keyword:`async`
  became a keyword in Python 3.7).

- Updated Python compatibility:

  - Python 3.6, 3.7 and 3.8 are now tested and supported.
  - Python 2.6 and 3.4 are no longer supported.

- Defined ``__all__`` for existing Python modules.

- Fixed :pypi:`humanfriendly` deprecation warnings and bumped dependencies that
  went through the same process.

- Fixed all :pypi:`Sphinx` broken references and configured Sphinx to report
  broken references as errors instead of warnings (to prevent broken references
  from piling up in the future).

- Various Travis CI fixes that didn't affect published code.

.. _Release 1.0: https://github.com/xolox/python-proc/compare/0.17...1.0
.. _pull request #2: https://github.com/xolox/python-proc/pull/2

`Release 0.17`_ (2018-06-27)
----------------------------

Added ``with-gui-environment`` program (a generalization of ``notify-send-headless``).

.. _Release 0.17: https://github.com/xolox/python-proc/compare/0.16...0.17

`Release 0.16`_ (2018-06-21)
----------------------------

Expose the value of ``/proc/[pid]/cwd`` on :class:`.Process` objects as the new
:attr:`.Process.cwd` property.

.. _Release 0.16: https://github.com/xolox/python-proc/compare/0.15...0.16

`Release 0.15`_ (2018-06-21)
----------------------------

- Changes related to the :mod:`proc.gpg` module and the ``with-gpg-agent`` program:

  - Use existing ``$GPG_AGENT_INFO`` values when available and validated.
  - Let the operator know when starting a new GPG agent daemon (through logging).
  - Check if :man:`gpg-agent`` is installed before trying to run ``gpg-agent --daemon``.
  - Added support for GPG agent sockets in ``/run/user/$UID`` (GnuPG >= 2.1.13).

    - This incompatibility came to light when I upgraded my laptop from Ubuntu
      16.04 to 18.04.

  - Fixed hanging Travis CI builds caused by ``gpg-agent --daemon`` not
    detaching properly when the standard error stream is redirected.

    - This incompatibility was exposed by Travis CI switching from Ubuntu 12.04
      to 14.04.

  - Fixed race condition in :func:`.find_gpg_agent_info()` raising
    :exc:`~exceptions.AttributeError`.

- Changes related to the documentation:

  - Added this change log to the documentation (with a link in the readme).
  - Integrated the :mod:`property_manager.sphinx` module (to generate
    boilerplate documentation).
  - Fixed intersphinx mapping in documentation configuration.
  - Changed HTML theme from `default` to `nature` (a wide layout).
  - Include documentation in source distributions (``MANIFEST.in``).

- And then some miscellaneous changes:

  - Fixed Apache WSGI configuration on Travis CI.

    - This test suite incompatibility was exposed by Travis CI switching from
      Ubuntu 12.04 to 14.04.

  - Restored Python 2.6 compatibility in the test suite (concerning
    :pypi:`pytest` version).
  - Added license=MIT key to ``setup.py`` script.
  - Bumped the copyright to 2018.

.. _Release 0.15: https://github.com/xolox/python-proc/compare/0.14...0.15

`Release 0.14`_ (2017-06-24)
----------------------------

Swallow exceptions in the :func:`.notify_desktop()` function.

This change is technically backwards incompatible but I consider it the more
sane behavior; I had just simply never seen ``notify-send`` fail until the
failure which prompted this release 😇.

.. _Release 0.14: https://github.com/xolox/python-proc/compare/0.13...0.14

`Release 0.13`_ (2017-06-24)
----------------------------

- Provide proper compatibility with GnuPG  >= 2.1 which uses the fixed
  location ``~/.gnupg/S.gpg-agent`` for the agent socket.
- Bug fix for systemd incompatibility in test suite.
- Moved test helpers to the :mod:`humanfriendly.testing` module.

.. _Release 0.13: https://github.com/xolox/python-proc/compare/0.12...0.13

`Release 0.12`_ (2017-02-14)
----------------------------

Improved robustness of Apache master process selection.

.. _Release 0.12: https://github.com/xolox/python-proc/compare/0.11...0.12

`Release 0.11`_ (2017-01-24)
----------------------------

Added ``with-gpg-agent`` program: A smart wrapper for the ``gpg-agent
--daemon`` functionality that makes sure the environment variable
``$GPG_AGENT_INFO`` is always set correctly.

.. _Release 0.11: https://github.com/xolox/python-proc/compare/0.10.1...0.11

`Release 0.10.1`_ (2016-11-13)
------------------------------

Fixed broken reStructuredText syntax in README (which breaks the rich text
rendering on the Python Package Index).

.. _Release 0.10.1: https://github.com/xolox/python-proc/compare/0.10...0.10.1

`Release 0.10`_ (2016-11-12)
----------------------------

- Several improvements to ``cron-graceful``:

  - Improved cron daemon termination.
  - Improved user friendliness of output.
  - Avoid useless log output noise.

- Start publishing wheel distributions.
- Explicitly signal skipped tests (when possible).
- Refactored internal project infrastructure such as the makefile, setup script
  and Travis CI build configuration.

.. _Release 0.10: https://github.com/xolox/python-proc/compare/0.9.1...0.10

`Release 0.9.1`_ (2016-06-13)
-----------------------------

Silenced another race condition (``ESRCH`` instead of ``ENOENT``).

This is one of those things that you only observe after running a package like
:pypi:`proc` from a periodic task (cron job) that runs every minute on a dozen

proc/gpg.py
# proc: Simple interface to Linux process information.
#
# Author: Peter Odding <peter@peterodding.com>
# Last Change: April 26, 2020
# URL: https://proc.readthedocs.io

"""
The :mod:`proc.gpg` module provides a smart wrapper for `gpg-agent --daemon`_.

.. contents::
   :local:

Introduction to gpg-agent
=========================

The gpg-agent is used to keep secret keys unlocked in between multiple
invocations of the gpg_ command to avoid retyping the same password. It's
usually started from a script that initializes your graphical session, ensuring
that all processes from that point onward inherit the environment variable
``$GPG_AGENT_INFO``. This variable together with the command line option ``gpg
--use-agent`` enable the use of the gpg-agent daemon.

.. note:: This applies to GnuPG versions before 2.1, refer to the `What's new
          in GnuPG 2.1 <https://www.gnupg.org/faq/whats-new-in-2.1.html#autostart>`_
          page for details. In GnuPG 2.1 the use of ``$GPG_AGENT_INFO`` was
          removed because it proved too cumbersome for users :-).

Problem statement
=================

Making sure that ``$GPG_AGENT_INFO`` is always set correctly can be a hassle.
For example I frequently use SSH_ to connect between my personal laptop and work
laptop and the interactive shells spawned by the SSH daemon have no relation to
any graphical session so they don't have ``$GPG_AGENT_INFO`` set.

Of course I can just execute ``eval $(gpg-agent --daemon)`` in an interactive
session to spawn a gpg-agent on the spot, but that will remain running in the
background indefinitely after I close the interactive session, without any
simple means of reconnecting.

Initial solution
================

Somewhere in 2016 I developed a Python script that used :mod:`proc.core` to
search for ``$GPG_AGENT_INFO`` values in ``/proc`` so I could easily reconnect
to previously spawned gpg-agents. It mostly worked but it could pick the wrong
``$GPG_AGENT_INFO`` when references remained to a crashed or killed agent, so
eventually I added checks that ensured the UNIX socket and agent process still
existed.

After using this for a while I discovered that when I started a new gpg-agent
from an interactive shell spawned by the SSH daemon, I would lose all means of
connecting to the agent as soon as I logged out of the interactive shell, even
though the agent remained running :-).

Revised solution
================

After taking a step back I realized that the problem could be approached from a
completely different angle: Why not search for an existing ``gpg-agent``
process and infer the required ``$GPG_AGENT_INFO`` value by inspecting the
process using lsof_?

The revised solution has worked quite well for me and so I'm now publishing it
as the :mod:`proc.gpg` module which implements the command line program
``with-gpg-agent``.

Internal documentation of :mod:`proc.gpg`
=========================================

.. _gpg-agent --daemon: https://manpages.debian.org/cgi-bin/man.cgi?query=gpg-agent
.. _gpg: https://manpages.debian.org/cgi-bin/man.cgi?query=gpg
.. _lsof: https://manpages.debian.org/cgi-bin/man.cgi?query=lsof
.. _SSH: https://manpages.debian.org/cgi-bin/man.cgi?query=ssh
"""

# Standard library modules.
import functools
import getopt
import os
import stat
import sys

# External dependencies.
import coloredlogs
from executor import ExternalCommandFailed, execute, which
from humanfriendly import Timer, parse_path
from humanfriendly.terminal import usage, warning
from humanfriendly.terminal.spinners import Spinner
from verboselogs import VerboseLogger

# Modules included in our package.
from proc.core import find_processes

# Public identifiers that require documentation.
__all__ = (
    'LAUNCH_TIMEOUT',
    'NEW_STYLE_SOCKET',
    'USAGE_TEXT',
    'enable_gpg_agent',
    'find_fixed_agent_socket',
    'find_gpg_agent_info',
    'find_open_unix_sockets',
    'get_gpg_variables',
    'have_agent_program',
    'have_valid_agent_info',
    'logger',
    'main',
    'parse_arguments',
    'start_gpg_agent',
    'validate_unix_socket',
    'with_gpg_agent',
)

LAUNCH_TIMEOUT = 30
"""
The timeout for a newly launched GPG agent daemon to come online (a number).

This gives the maximum number of seconds to wait for a newly launched GPG agent
daemon to come online.
"""

NEW_STYLE_SOCKET = '~/.gnupg/S.gpg-agent'
"""The location of the GPG agent socket for GnuPG 2.1 and newer (a string)."""

# Initialize a logger.
logger = VerboseLogger(__name__)

# Inject our logger into all execute() calls.
execute = functools.partial(execute, logger=logger)

USAGE_TEXT = """
Usage: with-gpg-agent [OPTIONS] COMMAND

Run the given COMMAND with the environment variable(s) required by gpg to
connect to a gpg-agent daemon. If no gpg-agent daemon is running yet a
new one will be spawned in the background.

Supported options:

  -v, --verbose

    Make more noise (increase verbosity).

  -q, --quiet

    Make less noise (decrease verbosity).

  -h, --help

    Show this message and exit.
"""


def main():
    """Wrapper for :func:`with_gpg_agent()` that feeds it :data:`sys.argv`."""
    coloredlogs.install(syslog=True)
    with_gpg_agent(sys.argv[1:])


def with_gpg_agent(arguments):
    """Command line interface for the ``with-gpg-agent`` program."""
    command_line = parse_arguments(arguments)
    try:
        execute(*command_line, environment=get_gpg_variables())
    except ExternalCommandFailed as e:
        logger.error(e.error_message)
        sys.exit(1)


def parse_arguments(arguments):
    """
    Parse the command line arguments.

    :param arguments: A list of strings with command line options and/or arguments.
    :returns: A list of strings with the positional arguments.
    """
    try:
        options, arguments = getopt.gnu_getopt(arguments, 'vqh', [
            'verbose', 'quiet', 'help'
        ])
        for option, value in options:
            if option in ('-v', '--verbose'):
                coloredlogs.increase_verbosity()
            elif option in ('-q', '--quiet'):
                coloredlogs.decrease_verbosity()
            elif option in ('-h', '--help'):
                usage(USAGE_TEXT)
                sys.exit(0)
            else:
                assert False, "Unhandled option!"
        if not arguments:
            usage(USAGE_TEXT)
            sys.exit(0)
        return arguments
    except Exception as e:
        warning("Error: Failed to parse command line arguments! (%s)", e)
        sys.exit(1)


def enable_gpg_agent(**options):
    """
    Update :data:`os.environ` with the variables collected by :func:`get_gpg_variables()`.

    :param options: Optional keyword arguments for :func:`get_gpg_variables()`.
    """
    os.environ.update(get_gpg_variables(**options))


def get_gpg_variables(timeout=LAUNCH_TIMEOUT):
    """
    Prepare the environment variable(s) required by the gpg_ program.

    :param timeout: The timeout for a newly launched GPG agent daemon to
                    start (a number, defaults to :data:`LAUNCH_TIMEOUT`).
    :returns: A dictionary with environment variables.

    This function tries to figure out the correct values of two
    environment variables that are used by the gpg_ program:

    - ``$GPG_AGENT_INFO`` is generated using :func:`find_gpg_agent_info()`,
      spawning a new agent if :func:`find_gpg_agent_info()` initially returns
      :data:`None`.

    - ``$GPG_TTY`` is generated using `/usr/bin/tty`_.

    .. _/usr/bin/tty: https://manpages.debian.org/cgi-bin/man.cgi?query=tty
    """
    environment = {}
    # Try to figure out the correct value of $GPG_AGENT_INFO.
    logger.debug("Preparing $GPG_AGENT_INFO variable ..")
    if have_valid_agent_info():
        logger.debug("Using existing value of $GPG_AGENT_INFO ..")
        environment['GPG_AGENT_INFO'] = os.environ['GPG_AGENT_INFO']
    else:
        gpg_agent_info = find_gpg_agent_info()
        if not gpg_agent_info:
            # Start a new GPG agent daemon?
            if have_agent_program():
                logger.debug("No running GPG agent found, trying to spawn new one ..")
                gpg_agent_info = start_gpg_agent(timeout=timeout)
            else:
                logger.notice("Not using gpg-agent program (it's not installed).")
        if gpg_agent_info:
            environment['GPG_AGENT_INFO'] = gpg_agent_info
    # Try to figure out the correct value of $GPG_TTY.
    logger.debug("Preparing $GPG_TTY variable ..")
    gpg_tty = execute('tty', capture=True, check=False, shell=False, tty=True)
    if gpg_tty:
        environment['GPG_TTY'] = gpg_tty
    logger.debug("GPG environment: %s", environment)
    return environment


def find_gpg_agent_info():
    """
    Reconstruct ``$GPG_AGENT_INFO`` based on a running ``gpg-agent`` process.

    :returns: A string or :data:`None`.

    This function uses :func:`~proc.core.find_processes()` to search for
    ``gpg-agent`` processes and runs lsof_ to find out which UNIX socket is
    being used by the agent. Based on this information it reconstructs
    the expected value of ``$GPG_AGENT_INFO``.
    """
    logger.debug("Searching for running GPG agent ..")
    our_uid = os.getuid()
    for process in find_processes():
        if process.exe_name == 'gpg-agent':
            logger.debug("Found GPG agent with PID %i, checking user id .. ", process.pid)
            their_uid = process.user_ids.real if process.user_ids else 'unknown'
            if our_uid == their_uid:
                socket_file = None
                logger.debug("GPG agent user id matches ours! Looking for UNIX socket ..")
                fixed_socket = find_fixed_agent_socket()
                if validate_unix_socket(fixed_socket):
                    logger.debug("Found GnuPG >= 2.1 compatible socket: %s", fixed_socket)
                    socket_file = fixed_socket
                else:
                    logger.debug("Using `lsof' to find open UNIX sockets ..")
                    for filename in find_open_unix_sockets(process.pid):
                        logger.debug("UNIX domain socket reported by lsof: %s", filename)
                        if validate_unix_socket(filename):
                            socket_file = filename
                            break
                # We will now reconstruct $GPG_AGENT_INFO based on the
                # information that we've gathered. We should end up with
                # an expression like `/tmp/gpg-KE5ZZL/S.gpg-agent:2407:1'.
                if socket_file:
                    agent_info = ':'.join([socket_file, str(process.pid), '1'])
                    logger.debug("Reconstructed $GPG_AGENT_INFO: %s", agent_info)
                    return agent_info
            else:
                logger.debug("GPG agent user id (%s) doesn't match ours (%i), ignoring process %i.",
                             their_uid, our_uid, process.pid)


def find_fixed_agent_socket():
    """
    Search for a GPG agent UNIX socket in one of the "fixed locations".

    :returns: The pathname of the found socket file (a string) or :data:`None`.

    Two locations are searched, in the given order (the first that is found is
    returned):

    - Starting from GnuPG 2.1.13 the location ``/run/user/$UID/gnupg/S.gpg-agent``
      is used (only when the directory ``/run/user/$UID`` exists).

    - GnuPG 2.1 removed the ``$GPG_AGENT_INFO`` related code and switched to
      the fixed location ``~/.gnupg/S.gpg-agent``.
    """
    socket_on_tmpfs = '/run/user/%i/gnupg/S.gpg-agent' % os.getuid()
    for socket in (socket_on_tmpfs, parse_path(NEW_STYLE_SOCKET)):
        if os.path.exists(socket):
            return socket


def find_open_unix_sockets(pid):
    """
    Find the pathnames of any UNIX sockets held open by a process.

    :param pid: The process ID (a number).
    :returns: A generator of pathnames (strings).
    """
    # A quick lsof tutorial :-)
    #  -F enables output that is easy to parse,
    #  -p lists the open files of a specific PID,
    #  -a combines -p and -U using AND instead of OR,
    #  -U lists only UNIX domain socket files.
    output = execute('lsof', '-F', '-p', str(pid), '-a', '-U', capture=True, check=False, silent=True)
    for line in output.splitlines():
        if line and line[0] == 'n':
            filename = line[1:]
            if filename:
                yield filename


def have_agent_program():
    """
    Check whether the ``gpg-agent`` program is installed.

    :returns: :data:`True` when the ``gpg-agent`` program is available on the
               ``$PATH``, :data:`False` otherwise.
    """
    return bool(which('gpg-agent'))


def have_valid_agent_info():
    """
    Check if the existing ``$GPG_AGENT_INFO`` value is usable.

    :returns: :data:`True` if the existing ``$GPG_AGENT_INFO`` is valid,
              :data:`False` otherwise.

    This function parses the ``$GPG_AGENT_INFO`` environment
    variable and validates the resulting UNIX socket filename
    using :func:`validate_unix_socket()`.
    """
    components = os.environ.get('GPG_AGENT_INFO', '').split(':')
    return components and validate_unix_socket(components[0])


def validate_unix_socket(pathname):
    """
    Check whether a filename points to a writable UNIX socket.

    :param pathname: The pathname of the socket file (a string).
    :returns: :data:`True` if the socket exists and is writable,
              :data:`False` in all other cases.
    """
    try:
        metadata = os.stat(pathname)
        if stat.S_ISSOCK(metadata.st_mode):
            return os.access(pathname, os.W_OK)
    except Exception:
        pass
    return False


def start_gpg_agent(timeout=LAUNCH_TIMEOUT):
    """
    Start a new gpg-agent daemon in the background.

    :param timeout: The timeout for the newly launched GPG agent daemon to
                    start (a number, defaults to :data:`LAUNCH_TIMEOUT`).
    :returns: The return value of :func:`find_gpg_agent_info()`.
    """
    timer = Timer()
    logger.info("Starting a new GPG agent daemon ..")
    execute('gpg-agent', '--daemon', asynchronous=True, silent=True)
    with Spinner(timer=timer) as spinner:
        while timer.elapsed_time < LAUNCH_TIMEOUT:
            gpg_agent_info = find_gpg_agent_info()
            if gpg_agent_info:
                logger.debug("Waited %s for GPG agent daemon to start.", timer)
                return gpg_agent_info
            spinner.step(label="Waiting for GPG agent daemon")
            spinner.sleep()
    logger.warning("Failed to locate spawned GPG agent! (waited for %s)", timer)
